pkg/apis/elasticsearch/v1/elasticsearch_types_test.go

```diff
@@ -314,3 +316,76 @@ func TestElasticsearch_DisabledPredicates(t *testing.T) {
 		})
 	}
 }
+
+// Test_AssociationConfs tests that if something resets the AssocConfs map, then AssociationConf() reinitializes
+// the map from the annotation.
+func Test_AssociationConfs(t *testing.T) {
+	// simple es without associations
+	es := &Elasticsearch{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "es",
+			Namespace: "default",
+		},
+	}
+
+	// set assoc conf even if no assoc
+	assert.Equal(t, 0, len(es.AssocConfs))
+	for _, association := range es.GetAssociations() {
+		assocConf, err := assocutils.GetAssociationConf(association)
+		assert.NoError(t, err)
+		association.SetAssociationConf(assocConf)
+	}
+	assert.Equal(t, 0, len(es.AssocConfs))
+
+	// checks that assocConfs is nil
+	for _, assoc := range es.GetAssociations() {
+		assert.Nil(t, assoc.AssociationConf())
+	}
+
+	// es with associations
+	esMon := &Elasticsearch{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "esmon",
+			Namespace: "default",
+			Annotations: map[string]string{
+				"association.k8s.elastic.co/es-conf-864518565":  `{"authSecretName":"es-default-metrics-beat-es-mon-user","authSecretKey":"default-es-default-esmon-beat-es-mon-user","caCertProvided":true,"caSecretName":"es-es-monitoring-default-metrics-ca","url":"https://metrics-es-http.default.svc:9200","version":"8.0.0"}`,
+				"association.k8s.elastic.co/es-conf-1654136115": `{"authSecretName":"es-default-logs-beat-es-mon-user","authSecretKey":"default-es-default-esmon-beat-es-mon-user","caCertProvided":true,"caSecretName":"es-es-monitoring-default-logs-ca","url":"https://logs-es-http.default.svc:9200","version":"8.0.0"}`,
+			},
+		},
+		Spec: ElasticsearchSpec{
+			Monitoring: Monitoring{
+				Metrics: MetricsMonitoring{
+					ElasticsearchRefs: []commonv1.ObjectSelector{{
+						Name:      "metrics",
+						Namespace: "default"},
+					},
+				},
+				Logs: LogsMonitoring{
+					ElasticsearchRefs: []commonv1.ObjectSelector{{
+						Name:      "logs",
+						Namespace: "default"},
+					},
+				},
+			},
+		},
+	}
+
+	// set assoc conf
+	assert.Equal(t, 0, len(esMon.AssocConfs))
+	for _, association := range esMon.GetAssociations() {
+		assocConf, err := assocutils.GetAssociationConf(association)
+		assert.NoError(t, err)
+		association.SetAssociationConf(assocConf)
+	}
+	assert.Equal(t, 2, len(esMon.AssocConfs))
+
+	// simulate the case where the assocConfs map is reset, which can happen if the resource is updated
```

Comment:
> simulate the case where the assocConfs map is reset, which can happen if the resource is updated

Sorry, I'm not sure to understand. `assocConfs` is not persisted, I think it is expected for it to be empty before [`FetchWithAssociations`](https://github.com/elastic/cloud-on-k8s/blob/721cccfd905f43c345c7957fbfe173ea3ebb6bbf/pkg/controller/association/conf.go#L50) kicks in ?

Reply:
The issue happens after [FetchWithAssociations](https://github.com/elastic/cloud-on-k8s/blob/721cccfd905f43c345c7957fbfe173ea3ebb6bbf/pkg/controller/association/conf.go#L50), when the resource is updated, it resets the map. This test shows that if it is reset after `SetAssociationConf()`, the map is repopulated on the fly by `AssociationConf()`. 
Maybe I should go further and delete `SetAssociationConf()` to avoid confusion. We could even delete the `FetchWithAssociations()` and just rely on `AssociationConf()` called in different places.

Reply:
> We could even delete the FetchWithAssociations() and just rely on AssociationConf() called in different places.

Thanks for the clarification ! My understanding is that it means that `FetchWithAssociations` is unreliable 😕 I wonder if it's really worth keeping it along with the unserialized annotations ? IIUC they are only here to spare a few CPU cycles by not using `json.Unmarshal` when the association config. is needed ?